File: pa6-cky-v3/java/src/nlpclass/ling/Tree.java

package nlpclass.ling;

//import nlpclass.util.Counter;
//import nlpclass.util.CollectionUtils;

import java.util.*;

/**
 * Represent linguistic trees, with each node consisting of a label
 * and a list of children.
 * @author Dan Klein
 */
public class Tree<L> {
    L label;
    List<Tree<L>> children;

    public List<Tree<L>> getChildren() {
        return children;
    }
    
    public void setChildren(List<Tree<L>> children) {
        this.children = children;
    }
    
    public L getLabel() {
        return label;
    }
    
    public void setLabel(L label) {
        this.label = label;
    }

    /* Returns true at the word(leaf) level of a tree */
    public boolean isLeaf() {
        return getChildren().isEmpty();
    }

    /* Returns true level of non-terminals which are directly above
     * single words(leafs) 
     */
    public boolean isPreTerminal() {
        return getChildren().size() == 1 && getChildren().get(0).isLeaf();
    }

    public boolean isPhrasal() {
        return !(isLeaf() || isPreTerminal());
    }

    /* Returns a list of words at the leafs of this tree gotten by
     * traversing from left to right 
     */
    public List<L> getYield() {
        List<L> yield = new ArrayList<L>();
        appendYield(this, yield);

File: pa6-cky-v3/python/ling/Tree.py

from Constituent import Constituent

class Tree:

    # TODO: deepCopy() ?

    def __init__(self, label, children = []):
        """
            The constructor.
        """
        self.label = label
        self.children = children

    def is_leaf(self):
        """
            Returns true at the word (leaf) level of a tree.
        """
        return len(self.children) == 0

    def is_preterminal(self):
        """
            Returns true level for non-terminals which are directly above
            single words (leaves).
        """
        return len(self.children) == 1 and self.children[0].is_leaf()

    def is_phrasal(self):
        # Seems to be defined incorrectly. Not used anywhere.
        return not self.is_leaf() and not self.is_preterminal()

    def _append_yield(self, leaf_labels):
        if self.is_leaf():
            leaf_labels.append(self.label)
            return
        for child in self.children:
            child._append_yield(leaf_labels)

    def get_yield(self):